cluster/certificates.go

```diff
@@ -0,0 +1,110 @@
+package cluster
+
+import (
+	"crypto/rsa"
+	"fmt"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/rancher/rke/hosts"
+	"github.com/rancher/rke/k8s"
+	"github.com/rancher/rke/pki"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/cert"
+)
+
+// SaveClusterCerts ...
+func SaveClusterCerts(kClient *kubernetes.Clientset, crts map[string]pki.CertificatePKI) error {
+	logrus.Infof("[certificates] Save kubernetes certificates as secrets")
+	for crtName, crt := range crts {
+		err := saveCrtToKubernetes(kClient, crtName, crt)
+		if err != nil {
+			return fmt.Errorf("Failed to save certificate [%s] to kubernetes: %v", crtName, err)
+		}
+	}
+	logrus.Infof("[certificates] Successfuly saved certificates as kubernetes secret [%s]", pki.CertificatesSecretName)
+	return nil
+}
+
+func GetClusterCerts(kClient *kubernetes.Clientset) (map[string]pki.CertificatePKI, error) {
+	logrus.Infof("[certificates] Getting Cluster certificates from Kubernetes")
+	certificatesNames := []string{
+		pki.CACertName,
+		pki.KubeAPICertName,
+		pki.KubeNodeName,
+		pki.KubeProxyName,
+		pki.KubeControllerName,
+		pki.KubeSchedulerName,
+	}
+	certMap := make(map[string]pki.CertificatePKI)
+	for _, certName := range certificatesNames {
+		secret, err := k8s.GetSecret(kClient, certName)
+		if err != nil {
+			return nil, err
+		}
+		secretCert, _ := cert.ParseCertsPEM(secret.Data["Certificate"])
+		secretKey, _ := cert.ParsePrivateKeyPEM(secret.Data["Key"])
+		secretConfig := string(secret.Data["Config"])
+		certMap[certName] = pki.CertificatePKI{
+			Certificate: secretCert[0],
+			Key:         secretKey.(*rsa.PrivateKey),
+			Config:      secretConfig,
+		}
+	}
+	logrus.Infof("[certificates] Successfully fetched Cluster certificates from Kubernetes")
+	return certMap, nil
+}
+
+func DeployClusterCerts(kubeHosts []hosts.Host, clusterCerts map[string]pki.CertificatePKI, forceDeploy bool) error {
+	logrus.Infof("[certificates] Deploying kubernetes certificates to Cluster nodes")
```

Comment:
We should consider that there can be diff auth providers, and may be:

* create a generic method, something like "saveConfigToNode".
* Deploying admin config will be generic
* Saving certs (or other type of auth) will be based on the auth provider chosen

Reply:
This generic method should be in a dedicated file; each auth component will be in its separate file.